code/modules/recycling/disposal/pipe.dm

```diff
@@ -0,0 +1,364 @@
+			return mask & (~secdir)
+
+/obj/structure/disposalpipe/junction/flip
+	icon_state = "pipe-j2"
+	initialize_dirs = DISP_DIR_LEFT | DISP_DIR_FLIP
+	flip_type = /obj/structure/disposalpipe/junction
+
+/obj/structure/disposalpipe/junction/yjunction
+	icon_state = "pipe-y"
+	initialize_dirs = DISP_DIR_LEFT | DISP_DIR_RIGHT
+	flip_type = null
+
+
+//a trunk joining to a disposal bin or outlet on the same turf
+/obj/structure/disposalpipe/trunk
+	icon_state = "pipe-t"
+	var/obj/linked 	// the linked obj/machinery/disposal or obj/disposaloutlet
+
+/obj/structure/disposalpipe/trunk/Initialize()
+	. = ..()
+	getlinked()
+
+/obj/structure/disposalpipe/trunk/Destroy()
+	if(linked)
+		if(istype(linked, /obj/structure/disposaloutlet))
+			var/obj/structure/disposaloutlet/D = linked
+			D.trunk = null
+		else if(istype(linked, /obj/machinery/disposal))
+			var/obj/machinery/disposal/D = linked
+			D.trunk = null
+	return ..()
+
+/obj/structure/disposalpipe/trunk/proc/getlinked()
+	linked = null
+	var/obj/machinery/disposal/D = locate() in src.loc
+	if(D)
+		linked = D
+		if (!D.trunk)
+			D.trunk = src
+
+	var/obj/structure/disposaloutlet/O = locate() in src.loc
+	if(O)
+		linked = O
+
+
+/obj/structure/disposalpipe/trunk/can_be_deconstructed(mob/user)
+	if(linked)
+		to_chat(user, "<span class='warning'>You need to deconstruct disposal machinery above this pipe!</span>")
+	else
+		. = 1
+
+// would transfer to next pipe segment, but we are in a trunk
+// if not entering from disposal bin,
+// transfer to linked object (outlet or bin)
+/obj/structure/disposalpipe/trunk/transfer(obj/structure/disposalholder/H)
+	if(H.dir == DOWN)		// we just entered from a disposer
+		return ..()		// so do base transfer proc
+	// otherwise, go to the linked object
+	if(linked)
+		var/obj/structure/disposaloutlet/O = linked
+		if(istype(O))
+			O.expel(H)	// expel at outlet
+		else
+			var/obj/machinery/disposal/D = linked
+			D.expel(H)	// expel at disposal
+	else
+		src.expel(H, get_turf(src), 0)	// expel at turf
+	return null
+
+/obj/structure/disposalpipe/trunk/nextdir(obj/structure/disposalholder/H)
+	if(H.dir == DOWN)
+		return dir
+	else
+		return 0
+
+// a broken pipe
+/obj/structure/disposalpipe/broken
+	desc = "A broken piece of disposal pipe."
+	icon_state = "pipe-b"
+	initialize_dirs = DISP_DIR_NONE
+	// broken pipes always have dpdir=0 so they're not found as 'real' pipes
+	// i.e. will be treated as an empty turf
+
+/obj/structure/disposalpipe/broken/deconstruct()
+	qdel(src)
+
+
+
+
+// called when movable is expelled from a disposal pipe or outlet
+// by default does nothing, override for special behaviour
+
+/atom/movable/proc/pipe_eject(direction)
+	return
+
+/obj/effect/decal/cleanable/blood/gibs/pipe_eject(direction)
+	var/list/dirs
+	if(direction)
+		dirs = list( direction, turn(direction, -45), turn(direction, 45))
+	else
+		dirs = GLOB.alldirs.Copy()
+
+	src.streak(dirs)
+
+/obj/effect/decal/cleanable/robot_debris/gib/pipe_eject(direction)
```

Comment:
this is copy-pasted code, can you unify it please?

Reply:
Not really, types are not the same.